File: java/664. Strange Printer.java
class Solution {
    public int strangePrinter(String s) {
        s = removeDuplicates(s);
        int n = s.length();
        Integer[][] cache = new Integer[n][n];
        return helper(0, s.length()-1, s, cache);
    }

    private int helper(int start, int end, String s, Integer[][] cache) {
        if (start > end) return 0;
        if (cache[start][end] != null) return cache[start][end];
        int answer = 1+helper(start+1, end, s, cache);
        for (int k = start+1; k <= end; k++) {
            if (s.charAt(start) == s.charAt(k)) {
                answer = Math.min(
                    answer,
                    helper(start, k-1, s, cache) + helper(k+1, end, s, cache)
                );
            }
        }
        return cache[start][end] = answer;
    }

    private String removeDuplicates(String s) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < s.length(); ) {
            char current = s.charAt(i);
            sb.append(current);
            while (i < s.length() && current == s.charAt(i)) {
                i++;
            }
        }
        return sb.toString();
    }
}

File: go/664. Strange Printer.go
func strangePrinter(s string) int {
    s = removeDuplicates(s)
    cache := make([][]int, len(s))
    for i := range cache {
        cache[i] = make([]int, len(s))
    }
    return helper(0, len(s)-1, s, cache)
}

func helper(start, end int, s string, cache [][]int) int {
    if start > end {
        return 0
    }
    if cache[start][end] > 0 {
        return cache[start][end]
    }
    answer := 1 + helper(start+1, end, s, cache)
    for i := start+1; i <= end; i++ {
        if s[i] == s[start] {
            answer = min(answer, helper(start, i-1, s, cache) + helper(i+1, end, s, cache))
        }
    }
    cache[start][end] = answer
    return answer
}

func removeDuplicates(s string) string {
    var sb strings.Builder
    for i := 0; i < len(s); {
        curr := s[i]
        sb.WriteByte(s[i])
        for i < len(s) && curr == s[i] {
            i++
        }
    }
    return sb.String()
}

